p2p/src/message_network.rs

```diff
@@ -86,6 +86,161 @@ impl_consensus_encoding!(
     relay
 );
 
+/// A bitcoin user agent defined by BIP-14. The user agent is sent in the version message when a
+/// connection between two peers is established. It is intended to advertise client software in a
+/// well-defined format.
+///
+/// ref: <https://github.com/bitcoin/bips/blob/master/bip-0014.mediawiki>
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct UserAgent {
+    user_agent: String,
+}
+
+impl_consensus_encoding!(UserAgent, user_agent);
+
+impl UserAgent {
+    /// Build a new user agent from the lowest level client software. For example: `Satoshi` is
+    /// used by Bitcoin Core.
+    ///
+    /// # Panics
+    ///
+    /// If the client name contains one of: `/ ( ) :`
+    pub fn new<S: AsRef<str>>(client_name: S, client_version: UserAgentVersion) -> Self {
+        let parsed_name = client_name.as_ref();
+        if parsed_name.chars().any(|c| matches!(c, '/' | '(' | ')' | ':')) {
+            panic!("client name cannot contain: / ( ) :");
+        }
+        let version = client_version.to_string();
+        let agent = format!("/{parsed_name}:{version}/");
+        Self { user_agent: agent }
+    }
+
+    /// Build a user agent, ignoring BIP-14 recommendations.
+    pub fn from_nonstandard<S: ToString>(agent: S) -> Self {
+        Self {
+            user_agent: agent.to_string()
+        }
+    }
```

Comment:
I _think_ this should take `String` because of C-CALLER-CONTROL from the API guidelines.

https://rust-lang.github.io/api-guidelines/flexibility.html#c-caller-control

Reply:
Users can pass `String`s to this if they want and there will be no efficiency loss. But the typical user will pass a `&'static str`.

Reply:
Schooled. Thanks.